apollo-router/src/plugins/telemetry/apollo_otlp_exporter.rs

```diff
@@ -0,0 +1,140 @@
+#[derivative(Debug)]
+pub(crate) struct ApolloOtlpExporter {
+    batch_config: BatchProcessorConfig,
+    endpoint: Url,
+    apollo_key: String,
+    resource_template: Resource,
+    intrumentation_library: InstrumentationLibrary,
+    #[derivative(Debug = "ignore")]
+    otlp_exporter: Arc<Mutex<opentelemetry_otlp::SpanExporter>>,
+}
+
+impl ApolloOtlpExporter {
+    pub(crate) fn new(
+        endpoint: &Url,
+        batch_config: &BatchProcessorConfig,
+        apollo_key: &str,
+        apollo_graph_ref: &str,
+        schema_id: &str,
+    ) -> Result<ApolloOtlpExporter, BoxError> {
+        tracing::debug!(endpoint = %endpoint, "creating Apollo OTLP traces exporter");
+
+        let mut metadata = MetadataMap::new();
+        metadata.insert("apollo.api.key", MetadataValue::try_from(apollo_key)?);
+
+        return Ok(Self {
+            endpoint: endpoint.clone(),
+            batch_config: batch_config.clone(),
+            apollo_key: apollo_key.to_string(),
+            resource_template: Resource::new([
+                KeyValue::new(
+                    "apollo.router.id",
+                    ROUTER_ID.get_or_init(Uuid::new_v4).to_string(),
+                ),
+                KeyValue::new("apollo.graph.ref", apollo_graph_ref.to_string()),
+                KeyValue::new("apollo.schema.id", schema_id.to_string()),
+                KeyValue::new(
+                    "apollo.user.agent",
+                    format!(
+                        "{}@{}",
+                        std::env!("CARGO_PKG_NAME"),
+                        std::env!("CARGO_PKG_VERSION")
+                    ),
+                ),
+                KeyValue::new("apollo.client.host", hostname()?),
+                KeyValue::new("apollo.client.uname", get_uname()?),
+            ]),
+            intrumentation_library: InstrumentationLibrary::new(
+                GLOBAL_TRACER_NAME,
+                Some(format!(
+                    "{}@{}",
+                    std::env!("CARGO_PKG_NAME"),
+                    std::env!("CARGO_PKG_VERSION")
+                )),
+                Option::<String>::None,
+                None,
+            ),
+            otlp_exporter: Arc::new(Mutex::new(
+                SpanExporterBuilder::from(
+                    opentelemetry_otlp::new_exporter()
+                        .tonic()
+                        .with_timeout(batch_config.max_export_timeout)
+                        .with_endpoint(endpoint.to_string())
+                        .with_metadata(metadata),
+                        // TBD(tim): figure out why compression seems to be turned off on our collector
+                        // .with_compression(opentelemetry_otlp::Compression::Gzip),
+                )
+                .build_span_exporter()?,
+            )),
+            // TBD(tim): do we need another batch processor for this?
+            // Seems like we've already set up a batcher earlier in the pipe but not quite sure.
```

Comment:
cc @BrynCooke you might be the best one to answer this

Reply:
I think we'll find out if we need this when we deploy internally.